### PR TITLE
cmake options for vectorization report and instruction set

### PR DESCRIPTION
**Pull request by XzzX**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,8 @@ set(CMAKE_CXX_EXTENSIONS OFF) #...without compiler extensions like gnu++11
 ########################################################################
 # User input options                                                   #
 ########################################################################
+option(ESPP_LOCAL_ARCHITECTURE "Use instruction set of the local architecture." OFF)
+option(ESPP_VEC_REPORT "Enable reporting of loop vectorization." OFF)
```
Comment (junghans) [inline CMakeLists.txt:26]:
```suggestion
option(ESPP_VEC_REPORT "Enable reporting of loop vectorization (only for Intel compiler)." OFF)
```

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -56,18 +56,41 @@ find_package(Threads REQUIRED)
...
+    endif()
+endif()
+if(CMAKE_CXX_COMPILER_ID MATCHES Clang)
+    if(ESPP_LOCAL_ARCHITECTURE)
+        target_compile_options(_espressopp PRIVATE "-march=native")
+    endif()
+endif()
+if(CMAKE_CXX_COMPILER_ID MATCHES Intel)
+    if(ESPP_LOCAL_ARCHITECTURE)
+        target_compile_options(_espressopp PRIVATE "-xHost")
```
Comment (jnvance) [inline src/CMakeLists.txt:73]: Can you also add `-restrict` here?

**Review by jnvance — CHANGES_REQUESTED** [review]

**Review by junghans — APPROVED** [review]

File: src/CMakeLists.txt (hunk 2 above)
Comment (XzzX) [reply src/CMakeLists.txt:73]: Interesting. Is my understanding correct that Intel uses this flag to enable checking for the `restrict` keyword? I hope it does not transform all pointers into `restrict`. And it only applies to `restrict`, `__restrict` is always enabled? Pretty weird,
https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/language-options/restrict-qrestrict.html

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -56,18 +56,41 @@ find_package(Threads REQUIRED)
 set(Boost_USE_MULTITHREADED ON)
 
 add_library(_espressopp ${ESPRESSO_SOURCES})
+
+# compiler specific configurations
+if(CMAKE_CXX_COMPILER_ID MATCHES GNU)
+    if(ESPP_LOCAL_ARCHITECTURE)
+        target_compile_options(_espressopp PRIVATE "-march=native")
```
Comment (junghans) [inline src/CMakeLists.txt:63]: Will it lead to problems if only the library is compiled with optimization flags, but not the executables (tests)?

Comment (XzzX) [reply src/CMakeLists.txt:63]: No, this does not change the ABI. It only tells the compiler which instruction set it is allowed to used.

File: src/CMakeLists.txt (hunk 2 above)
Comment (XzzX) [reply src/CMakeLists.txt]: added

Comment (jnvance) [reply src/CMakeLists.txt]: I think it's the latter. I use a lot of them in the pointer declarations in the vec submodule.

Comment (XzzX) [reply src/CMakeLists.txt:74]: My state is, I do think we do not need it but it does not hurt. So we can enable it.

**Review by jnvance — APPROVED** [review]

File: src/CMakeLists.txt (hunk 2 above)
Comment (jnvance) [reply src/CMakeLists.txt:74]: There is even `-qopt-zmm-usage=high` recommended for AVX-512 which slightly improves loop vectorization. But we can also decide to add it later.

**Pull request merged**